numba/np/arraymath.py

```diff
@@ -665,6 +666,136 @@ def _trivial_isnan(x):
         return _trivial_isnan
 
 
+@register_jitable
+def _impl_iscomplex_isreal_arr(fn, x):
+    ax = np.asarray(x) # NumPy uses asanyarray here!
+    return fn(ax.imag, 0)
+
+
+@register_jitable
+def np_iscomplex_isreal(x, operator_fn, opt_ret):
+
+    if type_can_asarray(x):
+        return lambda x: _impl_iscomplex_isreal_arr(operator_fn, x)
+    elif isinstance(x, types.Optional):
+        def impl(x):
+            if x is None:
+                return opt_ret
+            return _impl_iscomplex_isreal_arr(operator_fn, x)
+    else:
+        def impl(x):
+            return False
+    return impl
+
+
+@overload(np.iscomplex)
+def np_iscomplex(x):
+    return np_iscomplex_isreal(x, operator.ne, False)
+
+
+@overload(np.isreal)
+def np_isreal(x):
+    return np_iscomplex_isreal(x, operator.eq, True)
+
+
+@overload(np.iscomplexobj)
+def iscomplexobj(x):
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/type_check.py#L282-L320
+    val = x.key if isinstance(x, types.Optional) else x
+    dt = determine_dtype(val)
+    iscmplx = np.issubdtype(dt, np.complexfloating)
+    return lambda x: False if x is None else iscmplx
+
+
+@overload(np.isrealobj)
+def isrealobj(x):
+    # Return True if x is a not complex type or an array of complex numbers.
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/ccfbcc1cd9a4035a467f2e982a565ab27de25b6b/numpy/lib/type_check.py#L290-L322
+    def impl(x):
+        return not np.iscomplexobj(x)
+    return impl
+
+
+@overload(np.isscalar)
+def np_isscalar(num):
+    res = isinstance(num, (types.Number, types.UnicodeType, types.Boolean))
+
+    def impl(num):
+        return res
+    return impl
+
+
+@register_jitable
+def _within_tol(a, b, rtol, atol):
+    return np.less_equal(np.abs(a - b), atol + rtol * np.abs(b))
+
+
+@overload(np.isclose)
+def np_isclose(a, b, rtol=1e-05, atol=1e-08, equal_nan=False):
```

Comment:
Shouldn't the defaults here be types, not values?

Reply:
Hum...not sure. There are other overloaded functions that uses values, not types. See:
https://github.com/numba/numba/blob/13ece9b97e6f01f750e870347f231282325f60c3/numba/np/arraymath.py#L1416

Reply:
@stuartarchibald, mind giving a second opinion here?

Reply:
Custom is values.

Reply:
The comment got lost in a review, https://github.com/numba/numba/pull/4610#discussion_r477271713 but custom is values.